File: contract-engine/app/template_requirements.py
from __future__ import annotations

from typing import Any, Dict, List


TEMPLATE_REQUIREMENTS: dict[str, list[tuple[str, str]]] = {
    "asset_purchase_agreement": [
        ("business_name", "Business Name"),
        ("seller_name", "Seller Name"),
        ("buyer_name", "Buyer Name"),
        ("agreement_date", "Agreement Date"),
        ("closing_date", "Closing Date"),
        ("purchase_price", "Purchase Price"),
        ("state", "State"),
    ],
    "bill_of_sale": [
        ("business_name", "Business Name"),
        ("seller_name", "Seller Name"),
        ("buyer_name", "Buyer Name"),
        ("seller_address", "Seller Address"),
        ("buyer_address", "Buyer Address"),
        ("closing_date", "Closing Date"),
        ("purchase_price", "Purchase Price"),
        ("state", "State"),
    ],
    "promissory_note": [
        ("business_name", "Business Name"),
        ("seller_name", "Seller Name"),
        ("buyer_name", "Buyer Name"),
        ("state", "State"),
    ],
    "non_compete": [
        ("business_name", "Business Name"),
        ("seller_name", "Seller Name"),
        ("buyer_name", "Buyer Name"),
        ("closing_date", "Closing Date"),
        ("state", "State"),
        ("non_compete_years", "Non-Compete Years"),
        ("non_compete_miles", "Non-Compete Miles"),
    ],
}


def _is_missing(value: Any) -> bool:
    if value is None:
        return True

    if isinstance(value, str):
        return value.strip() == ""

    return False


def validate_template_payload(
    template_key: str,
    data: Dict[str, Any],
) -> List[Dict[str, str]]:
    requirements = TEMPLATE_REQUIREMENTS.get(template_key, [])
    missing: List[Dict[str, str]] = []

    for field_key, label in requirements:
        if _is_missing(data.get(field_key)):
            missing.append(
                {
                    "field": field_key,
                    "label": label,
                }
            )

    return missing


def validate_payloads(
    payloads: list[Any],
) -> List[Dict[str, Any]]:
    validation_errors: List[Dict[str, Any]] = []

    for payload in payloads:
        template_key = payload.templateKey
        missing = validate_template_payload(template_key, payload.data)

        if missing:
            validation_errors.append(
                {
                    "template_key": template_key,
                    "document_name": payload.documentName,
                    "missing_fields": missing,
                }
            )

    return validation_errors

File: lib/template-requirements.ts
export type TemplateKey =
  | "asset_purchase_agreement"
  | "bill_of_sale"
  | "promissory_note"
  | "non_compete"
  | "irs_8594";

export type RequirementField = {
  field: string;
  label: string;
  treatZeroAsMissing?: boolean;
};

export const TEMPLATE_REQUIREMENTS: Record<TemplateKey, RequirementField[]> = {
  asset_purchase_agreement: [
    { field: "business_name", label: "Business Name" },
    { field: "seller_name", label: "Seller Name" },
    { field: "buyer_name", label: "Buyer Name" },
    { field: "agreement_date", label: "Agreement Date" },
    { field: "closing_date", label: "Closing Date" },
    { field: "purchase_price", label: "Purchase Price" },
    { field: "included_assets_text", label: "Included Assets" },
    { field: "excluded_assets_text", label: "Excluded Assets" },
    { field: "assumed_liabilities_text", label: "Assumed Liabilities" },
    { field: "excluded_liabilities_text", label: "Excluded Liabilities" },
    { field: "allocation_total", label: "Allocation Total", treatZeroAsMissing: true },
    { field: "state", label: "State" },
    { field: "included_assets_text", label: "Included Assets" },
    { field: "excluded_assets_text", label: "Excluded Assets" },
    { field: "assumed_liabilities_text", label: "Assumed Liabilities" },
    { field: "excluded_liabilities_text", label: "Excluded Liabilities" },
  ],
  bill_of_sale: [
    { field: "business_name", label: "Business Name" },
    { field: "seller_name", label: "Seller Name" },
    { field: "buyer_name", label: "Buyer Name" },
    { field: "seller_address", label: "Seller Address" },
    { field: "buyer_address", label: "Buyer Address" },
    { field: "closing_date", label: "Closing Date" },
    { field: "purchase_price", label: "Purchase Price" },
    { field: "state", label: "State" },
  ],
  promissory_note: [
    { field: "business_name", label: "Business Name" },
    { field: "seller_name", label: "Lender / Seller Name" },
    { field: "buyer_name", label: "Borrower / Buyer Name" },
    { field: "state", label: "State" },
    { field: "seller_financing_amount", label: "Seller Financing Amount" },
    { field: "closing_date", label: "Issue Date / Closing Date" },
    { field: "seller_financing_amount", label: "Seller Financing Amount", treatZeroAsMissing: true },
    { field: "promissory_interest_rate", label: "Interest Rate" },
    { field: "promissory_term_months", label: "Term Months" },
    { field: "promissory_first_payment_date", label: "First Payment Date" },
    { field: "promissory_maturity_date", label: "Maturity Date" },
    { field: "promissory_interest_rate", label: "Interest Rate" },
    { field: "promissory_term_months", label: "Term Months" },
    { field: "promissory_first_payment_date", label: "First Payment Date" },
    { field: "promissory_maturity_date", label: "Maturity Date" },
  ],
  non_compete: [
    { field: "business_name", label: "Business Name" },
    { field: "seller_name", label: "Restricted Party / Seller Name" },
    { field: "buyer_name", label: "Protected Party / Buyer Name" },
    { field: "closing_date", label: "Closing Date" },
    { field: "state", label: "State" },
    {
      field: "non_compete_years",
      label: "Non-Compete Years",
      treatZeroAsMissing: true,
    },
    {
      field: "non_compete_miles",
      label: "Non-Compete Miles",
      treatZeroAsMissing: true,
    },
    { field: "allocated_non_compete", label: "Non-Compete Allocation", treatZeroAsMissing: true },
    { field: "non_compete_restricted_business", label: "Restricted Business" },
    { field: "non_compete_territory", label: "Restricted Territory" },
    { field: "non_compete_restricted_business", label: "Restricted Business" },
    { field: "non_compete_territory", label: "Restricted Territory" },
  ],
  irs_8594: [
    { field: "business_name", label: "Business Name" },
    { field: "seller_name", label: "Seller Name" },
    { field: "buyer_name", label: "Buyer Name" },
    { field: "closing_date", label: "Closing Date" },
    { field: "purchase_price", label: "Purchase Price", treatZeroAsMissing: true },
    { field: "allocated_inventory", label: "Allocated Inventory" },
    { field: "allocated_ffe", label: "Allocated FFE" },
    {
      field: "allocated_non_compete",
      label: "Allocated Non-Compete",
      treatZeroAsMissing: true,
    },
    { field: "allocated_goodwill", label: "Allocated Goodwill" },
    { field: "allocated_leasehold", label: "Allocated Leasehold" },
    { field: "allocated_customer_contracts", label: "Allocated Customer Contracts" },
    { field: "allocated_trade_name", label: "Allocated Trade Name / Website / Phone Numbers" },
  ],
};

function isMissingValue(value: unknown, treatZeroAsMissing = false) {
  if (value === null || value === undefined) return true;

  if (typeof value === "string") {
    return value.trim() === "";
  }

  if (typeof value === "number") {
    if (Number.isNaN(value)) return true;
    if (treatZeroAsMissing && value <= 0) return true;
    return false;
  }

  return false;
}

export function validateTemplateData(
  templateKey: TemplateKey,
  data: Record<string, unknown>
) {
  const requirements = TEMPLATE_REQUIREMENTS[templateKey] ?? [];

  const missingFields = requirements.filter((item) =>
    isMissingValue(data[item.field], item.treatZeroAsMissing)
  );

  return {
    template_key: templateKey,
    missing_fields: missingFields,
    is_ready: missingFields.length === 0,
  };
}